### ISSUE TITLE
Should we unit test public static final fields of a class?

### ISSUE THREAD

**Issue by Cheng20010201** [nus-cs2103-AY2021S2/forum#44]
e.g.
```
public class Dummy {
    public static final String DESCRIPTION = "This is a dummy class.";
}
```
Should we unit test such fields? Such testing seems to be tedious and a bit ... meaningless?
If yes, what are the best practices?

Thanks!

**Comment by tjtanjin** [comment]
I have yet to construct unit test for such fields and I do agree it seems quite meaningless xD However, should there be a situation where there is a need to ensure the description is correct, I would likely construct a test case as below within the DummyTest class:

```
@Test
void checkDescription_whenInputDescription_thenDescriptionMatch() {
    String expected = "This is a dummy class.";
    Dummy dummy = new Dummy();
    assertEquals(expected, dummy.DESCRIPTION);
}
```

**Comment by Cheng20010201** [comment]
> I have yet to construct unit test for such fields and I do agree it seems quite meaningless xD However, should there be a situation where there is a need to ensure the description is correct, I would likely construct a test case as below within the DummyTest class:
> 
> ```
> @Test
> void checkDescription_whenInputDescription_thenDescriptionMatch() {
>     String expected = "This is a dummy class.";
>     Dummy dummy = new Dummy();
>     assertEquals(expected, dummy.DESCRIPTION);
> }
> ```

Agreed. If I am not wrong you are saying there would be cases where there is a need to check whether the description **matches** the expected value, in this case, the String `"This is a dummy class."`. 

Just want to extend the discussion a little bit: normally we would have a lot of such static final field declarations in our application, then managing the 'expectations' would be quite troublesome already. I guess there would be some place/some way for us to manage what would be expected for the (important) static fields easily (e.g. instead of peeking at individual classes), but I have no clue : |. Anyone knows anything about it?

**Comment by tjtanjin** [comment]
> Just want to extend the discussion a little bit: normally we would have a lot of such static final field declarations in our application, then managing the 'expectations' would be quite troublesome already. I guess there would be some place/some way for us to manage what would be expected for the (important) static fields easily (e.g. instead of peeking at individual classes), but I have no clue : |. Anyone knows anything about it?

I agree it's going to be huge hassle with lots of static final field declarations. Off the top of my head, I imagine for larger applications with lots of static final field declarations of string types, these strings may be stored/read from inside a `messages.yml` or `messages.json` file of sorts (has the added benefit of easy configuration/changing messages without directly editing the file containing code). It might then be possible to construct a single test class pertaining to all these messages which should at least centralise the testing and makes things slightly more organised. What do you think?

**Comment by Cheng20010201** [comment]
> I agree it's going to be huge hassle with lots of static final field declarations. Off the top of my head, I imagine for larger applications with lots of static final field declarations of string types, these strings may be stored/read from inside a `messages.yml` or `messages.json` file of sorts (has the added benefit of easy configuration/changing messages without directly editing the file containing code). It might then be possible to construct a single test class pertaining to all these messages which should at least centralise the testing and makes things slightly more organised. What do you think?

I used to think of putting important static messages in a doc or something equivelent. But yeah, using a `.json` or `.yml` file as a whole seems to be a much better idea. :D

**Comment by rajobasu** [comment]
Um, I have a question that how do you exactly define ```static final``` variables from an external file? The only way I can think of is using constructs like 
```
class A{
    static {
    
    }
}
```

However, is this a good way to do this? What if the thing fails? Exception handling will be hard as well since you have to do everything mostly in these places, so not that modular right? 

Just want to know what the best practices are for these things.

**Comment by tjtanjin** [comment]
@rajobasu By external file do you mean other java files within the same package? I imagine static variables in these files would then be referred to through their class names so maybe something like `Dummy.staticVariableHere` (static methods would also be referenced similarly).

**Comment by rajobasu** [comment]
No I mean, how you were talking about keeping static constants in a ```.json``` or something. You will have to load them right?

**Comment by tjtanjin** [comment]
@rajobasu Both [JSON](https://www.w3schools.com/whatis/whatis_json.asp) and [YAML](https://circleci.com/blog/what-is-yaml-a-beginner-s-guide/) are file formats for storing data. You are right to say they will have to be loaded and these are usually done through third-party libraries which differ across programming languages. For JSON in java, we have the likes of [Jackson](https://github.com/FasterXML/jackson) and [json-simple](https://code.google.com/archive/p/json-simple/). These libraries provide means for us to load in the data from those files which we can then assign to `static final` variables. Hope that clarifies :D

Edit: Exception handling will need to be done for reading/writing from `.json` and `.yml` files but is handled within the code. The files themselves just hold the information. I am not actually sure this is the best way to do things but there is the benefit of consolidating messages within a file and modifying them without having to touch the codebase directly.

**Comment by rajobasu** [comment]
Yea I know about JSON/json loaders but the thing is that where will you put the code for that? since it’s a static final variable, needs to be in static blocks right? But then, assuming all classes usually have some static constants, you can’t really instantiate anything in that case right due to circular logic. So you would end up having to put a lot of the initialisation code inside the static blocks, which don’t feel ideal to me. Agreed that there is a benefit of keeping constants at the same place, but do you even really want that? 

Think about it from another perspective :  not all constants are ```public``` (and in general most aren't). Putting them in the same place can be dangerous. Secondly, if you want to keep separate files for the constants in different class, might as well just hardcode them. 
Atleast these came to my mind!

**Comment by tjtanjin** [comment]
@rajobasu That's a really great point. I have been theory-crafting up till this point and am honestly unsure of how file contents can be specifically loaded into `static final` variables. I did a quick google search and you are right about the approach with initialisation class/codes. As to why they would not feel ideal, could you share more? I imagine even if this doesn't work well for messages, would it still at least work well for a configuration file?

Edit: I think it might be possible to group constants of a similar nature together instead of by class (for example messages pertaining to information). I am not sure what associated dangers there may be because any sensitive information should not be stored in plain text to begin with. Are there any dangers I am overlooking?
